platform/openide.util/src/org/openide/util/NbBundle.java

```diff
@@ -1458,4 +1475,90 @@ public DebugInputStream(InputStream base, int id, boolean localizable) {
 
         }
     }
+    
+    
+    /**
+     * Local charset to decode using UTF-8 by default, but automatically switching to ISO-8859-1 if UTF-8 decoding fails.
+     * 
+     */
+    static private class UtfThenIsoCharset extends Charset {
+
+        private final boolean onlyUTF8;
+
+        /**
+         *
+         * @param acceptOnlyUTF8 If true there is no automatic switch to ISO-8859-1 if UTF-8 decoding fails.
+         */
+        public UtfThenIsoCharset(boolean acceptOnlyUTF8) {
+            super(UtfThenIsoCharset.class.getCanonicalName(), null);
+            this.onlyUTF8 = acceptOnlyUTF8;
+        }
+
+        @Override
+        public boolean contains(Charset arg0) {
+            return false;
+        }
+
+        @Override
+        public CharsetDecoder newDecoder() {
+            return new UtfThenIsoDecoder(this, 1.0f, 1.0f);
+        }
+
+        @Override
+        public CharsetEncoder newEncoder() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+
+        private final class UtfThenIsoDecoder extends CharsetDecoder {
```

Comment:
Sorry I don't understand what you mean. I will make the 2 `Utf8ThenIsoCharset ` instances static. But the `UtfThenIsoDecoder` instances remain created on demand via Charset.newDecoder() and discarded right after loading, so I don't see the need to make `UtfThenIsoDecoder `stateful

Note: my code is doing exactly the same as JDK's code...

Reply:
For the note: I hope you don't imply there is code in this, that was copied from a foreign source (aka JDK). This would be invalid and would violate the assumption, that you are the author of the code.

For my comment, please have a look at the top of the https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CharsetDecoder.html. There the sequence is described, which methods should be invoked by callers. The ISO-8859-1 and UTF-8 decoders might not have internal state, in that case your overriders are ok, but if they have, they are invalid according to the described call sequence.

Reply:
No don't worry I'm the author of the code, I have not simply copied/paste from JDK. But the CharsetDecoder algorithm used from JDK9 is so obvious and simple (3 lines: decode with UTF, if error then rewind and switch to ISO), that I can guarantee my CharsetDecoder will behave the same.

Reply:
I checked the CharsetDecoder API and I see is no inconsistency with the code: you can switch decoder once on the ByteBuffer as long as you do it before the end of input.